akka-docs/rst/scala/persistence-schema-evolution.rst

```diff
@@ -0,0 +1,386 @@
+
+Detach domain model from data model
+-----------------------------------
+
+**Situation:**
+You want to separate the application model (often called the "*domain model*") completely from the models used to
+persist the corresponding events (the "*data model*"). For example because the data representation may change
+independently of the domain model.
+
+Another situation where this technique may be useful is when your serialization tool of choice requires generated
+classes to be used for serialization and deserialization of objects, like for example `Google Protocol Buffers`_ do,
+yet you do not want to leak this implementation detail into the domain model itself, which you'd like to model as
+plain Scala case classes.
+
+**Solution:**
+In order to detach the domain model, which is often represented using pure scala (case) classes, from the data model
+classes which very often may be less user-friendly yet highly optimised for throughput and schema evolution
+(like the classes generated by protobuf for example), it is possible to use a simple EventAdapter which maps between
+these types in a 1:1 style as illustrated below:
+
+.. figure:: ../images/persistence-detach-models.png
+   :align: center
+
+   Domain events (**A**) are adapted to the data model events (**D**) by the ``EventAdapter``.
+   The data model can be a format natively understood by the journal, such that it can store it more efficiently or
+   include additional data for the event (e.g. tags), for ease of later querying.
+
+We will use the following domain and data models to showcase how the separation can be implemented by the adapter:
+
+.. includecode:: code/docs/persistence/PersistenceSchemaEvolutionDocSpec.scala#detach-models
+
+The :class:`EventAdapter` takes care of converting from one model to the other one (in both directions),
+alowing the models to be completely detached from each other, such that they can be optimised independently
+as long as the mapping logic is able to convert between them:
+
+.. includecode:: code/docs/persistence/PersistenceSchemaEvolutionDocSpec.scala#detach-models-adapter
+
+**Additional technique - human readable events using "pass-through" journals:**
+
+The above described technique of detaching the domain model from the data model can be used in interesting ways
+when the journal can be smart about the type of events it receives. For example, let us consider a fictional journal
+plugin implementation that is able to receive ``JSON`` objects, and store them natively in the underlying datastore.
+An example of a datastore which may well implement this pattern is MongoDB, however other databases such as PostgreSQL
+and Cassandra could also do it because of their built-in JSON capabilities.
+
+In this aproach, the :class:`EventAdapter` is used as the marshalling layer: it serializes the events to/from JSON.
+The journal plugin notices that the incoming event type is JSON (for example by performing a ``match`` on the incoming
+event) and
```

Comment:
The whole section feels very incomplete. It needs more than the rest of that sentence I think.